### PR TITLE
feat: [CDS-81574]: Adding Helm Steady state check as account settings.

### PR DESCRIPTION
**Pull request by GoelPratyush99**
## Describe your changes
Adding Helm Steady state check as account settings.
Screenshot 2023-11-08 at 7 44 18 PM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55181)

### CODE REVIEW

**Comment by GoelPratyush99** [comment]
trigger uts

File: 120-ng-manager/modules/ng-settings/src/main/resources/io/harness/ngsettings/settings.yml [hunk 1]
```diff
@@ -540,5 +540,13 @@ settings :
     defaultValue: false
     valueType: BOOLEAN
     allowOverrides: false
+    allowedScopes:
+      - ACCOUNT
+  - identifier: enable_steady_state_for_jobs_native_helm
```
Comment (abosii-harness) [inline 120-ng-manager/modules/ng-settings/src/main/resources/io/harness/ngsettings/settings.yml:545]: should be this instead native_helm_enable_steady_state_for_jobs

File: 120-ng-manager/modules/ng-settings/src/main/resources/io/harness/ngsettings/settings.yml [hunk 2]
```diff
@@ -540,5 +540,13 @@ settings :
...
+    name: Enable Steady State For Jobs Native Helm
+    category: CD
+    defaultValue: false
+    valueType: BOOLEAN
+    allowOverrides: false
     allowedScopes:
```
Comment (abosii-harness) [inline 120-ng-manager/modules/ng-settings/src/main/resources/io/harness/ngsettings/settings.yml:551]: Don't we want to be configurable per project?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/helm/HelmDeployStep.java [hunk 3]
```diff
@@ -273,8 +289,7 @@ public TaskChainResponse executeHelmTask(ManifestOutcome manifestOutcome, Ambian
             .helmVersion(nativeHelmStepHelper.getHelmVersionBasedOnFF(
                 helmChartManifestOutcome.getHelmVersion(), AmbianceUtils.getAccountId(ambiance)))
             .namespace(cdStepHelper.getK8sInfraDelegateConfig(infrastructure, ambiance).getNamespace())
-            .k8SteadyStateCheckEnabled(cdFeatureFlagHelper.isEnabled(
-                AmbianceUtils.getAccountId(ambiance), FeatureName.HELM_STEADY_STATE_CHECK_1_16))
+            .k8SteadyStateCheckEnabled(ENABLE_STEADY_STATE_FOR_JOBS_TRUE_VALUE.equals(steadyStateForJobsSettingValue))
```
Comment (abosii-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/helm/HelmDeployStep.java:292]: The settings shouldn't change the behaviour of the flag but instead should be a different parameter that will skip the jobs steady state check, I believe this should be assigned with V2 FF

**Review by abosii-harness — CHANGES_REQUESTED** [review]

File: 120-ng-manager/modules/ng-settings/src/main/resources/io/harness/ngsettings/settings.yml (hunk 1 above)
Comment (GoelPratyush99) [reply 120-ng-manager/modules/ng-settings/src/main/resources/io/harness/ngsettings/settings.yml]: done

File: 125-cd-nextgen/src/main/java/io/harness/cdng/helm/HelmDeployStep.java (hunk 3 above)
Comment (GoelPratyush99) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/helm/HelmDeployStep.java]: done

File: 120-ng-manager/modules/ng-settings/src/main/resources/io/harness/ngsettings/settings.yml (hunk 2 above)
Comment (GoelPratyush99) [reply 120-ng-manager/modules/ng-settings/src/main/resources/io/harness/ngsettings/settings.yml:551]: done

**Comment by GoelPratyush99** [comment]
trigger uts

File: 120-ng-manager/modules/ng-settings/src/main/resources/io/harness/ngsettings/settings.yml [hunk 4]
```diff
@@ -541,4 +541,14 @@ settings :
     valueType: BOOLEAN
     allowOverrides: false
     allowedScopes:
-      - ACCOUNT
\ No newline at end of file
+      - ACCOUNT
+  - identifier: native_helm_enable_steady_state_for_jobs
+    name: Enable Steady State For Jobs In Native Helm Deployment
```
Comment (abhinavsingh1196) [inline 120-ng-manager/modules/ng-settings/src/main/resources/io/harness/ngsettings/settings.yml:546]: nit: Should this be
Perform steady state **check** for Jobs in Native Helm deployments
?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/helm/HelmRollbackStep.java [hunk 5]
```diff
@@ -207,6 +212,18 @@ public TaskRequest obtainTaskAfterRbac(
         CDStepHelper.getParameterFieldBooleanValue(helmRollbackStepParams.getSkipSteadyStateCheck(),
             HelmRollbackBaseStepInfoKeys.skipSteadyStateCheck, stepParameters);
 
+    String steadyStateForJobsSettingValue = "";
```
Comment (abhinavsingh1196) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/helm/HelmRollbackStep.java:215]: duplicated code in both helm steps, good candidate to extract it. Also, we can include the below equality check in that method too and return boolean `enableSteadyStateCheckForJobs` from it

File: 951-delegate-task-commons/src/main/java/software/wings/beans/TaskType.java [hunk 6]
```diff
@@ -534,7 +534,8 @@ public enum TaskType {
   AWS_ASG_SHIFT_TRAFFIC_TASK_NG(TaskGroup.ASG, "AWS Asg Shift Traffic"),
   AWS_ASG_BLUE_GREEN_ROLLBACK_TASK_NG_V3(TaskGroup.ASG, "AWS Asg Blue Green Rollback V3 with Shift Traffic support"),
   AWS_ASG_BLUE_GREEN_PREPARE_ROLLBACK_DATA_TASK_NG_V3(
-      TaskGroup.ASG, "AWS Asg Blue Green Prepare Rollback Data for Shift Traffic");
+      TaskGroup.ASG, "AWS Asg Blue Green Prepare Rollback Data for Shift Traffic"),
+  HELM_COMMAND_TASK_NG_JOBS_STEADY_STATE_CHECK(TaskGroup.HELM, "Helm Command Task with Oci Helm Ecr Config");
```
Comment (abhinavsingh1196) [inline 951-delegate-task-commons/src/main/java/software/wings/beans/TaskType.java:538]: Wrong task description

**Review by abhinavsingh1196 — CHANGES_REQUESTED** [review]

File: 951-delegate-task-commons/src/main/java/software/wings/beans/TaskType.java (hunk 6 above)
Comment (GoelPratyush99) [reply 951-delegate-task-commons/src/main/java/software/wings/beans/TaskType.java:538]: done

File: 125-cd-nextgen/src/main/java/io/harness/cdng/helm/HelmRollbackStep.java (hunk 5 above)
Comment (GoelPratyush99) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/helm/HelmRollbackStep.java:215]: done

File: 120-ng-manager/modules/ng-settings/src/main/resources/io/harness/ngsettings/settings.yml (hunk 4 above)
Comment (GoelPratyush99) [reply 120-ng-manager/modules/ng-settings/src/main/resources/io/harness/ngsettings/settings.yml:546]: done

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger uts

**Review by acasian-harness — COMMENTED** [review]
@GoelPratyush99 have we validated the override part: If the setting is at project level true but false for org and account level - is this considering enabled?

**Review by GoelPratyush99 — COMMENTED** [review]
Hey @acasian-harness! I have verified the following scenario. I am not sure what overrides does so I am leaving it as true as its true for others having all the scope. 

project               org                   account               allow overrides              result
true                    false                 false                    false                                 true
true                    false                 false                    true                                  true













> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55181#-:-Nio-Fag9mgQhNh8YOaI:b-yt7e08)* status: 18 file(s) to review. 2 change requests. 5. unresolved discussions. 0 of 2 approvals obtained.

**Review by nishantsaini20 — APPROVED** [review]
Approved for settings.yml

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
**None of the UT failure are associated with my changes**

File: 125-cd-nextgen/src/test/java/io/harness/cdng/helm/HelmRollbackStepTest.java [hunk 7]
```diff
@@ -149,7 +149,28 @@ public void testRollback() {
             .output(NativeHelmDeployOutcome.builder().hasInstallUpgradeStarted(true).releaseName("test").build())
             .build();
 
-    testRollback(releaseOutput, deploymentOutput, "test");
+    HelmRollbackCommandRequestNG requestNG = testRollback(releaseOutput, deploymentOutput, "test");
+    assertThat(requestNG.isUseSteadyStateCheckForJobs()).isFalse();
+  }
+
+  @Test
+  @Owner(developers = ABOSII)
```
Comment (abhinavsingh1196) [inline 125-cd-nextgen/src/test/java/io/harness/cdng/helm/HelmRollbackStepTest.java:157]: wrong owner

Comment (TarunUba1) [inline 125-cd-nextgen/src/test/java/io/harness/cdng/helm/HelmRollbackStepTest.java:157]: You are the owner here

**Review by abhinavsingh1196 — DISMISSED** [review]

**Review by TarunUba1 — DISMISSED** [review]

File: 125-cd-nextgen/src/test/java/io/harness/cdng/helm/HelmRollbackStepTest.java (hunk 7 above)
Comment (TarunUba1) [reply 125-cd-nextgen/src/test/java/io/harness/cdng/helm/HelmRollbackStepTest.java:157]: copying my comment

**Review by abosii-harness — CHANGES_REQUESTED** [review]

File: 120-ng-manager/modules/ng-settings/src/main/resources/io/harness/ngsettings/settings.yml [hunk 8]
```diff
@@ -541,4 +541,14 @@ settings :
...
     allowedScopes:
-      - ACCOUNT
\ No newline at end of file
+      - ACCOUNT
+  - identifier: native_helm_enable_steady_state_for_jobs
+    name: Perform steady state check for Jobs in Native Helm deployments
+    category: CD
+    defaultValue: false
+    valueType: BOOLEAN
+    allowOverrides: false
```
[the remark this answers is not in the document]
Comment (GoelPratyush99) [reply 120-ng-manager/modules/ng-settings/src/main/resources/io/harness/ngsettings/settings.yml]: done

File: 125-cd-nextgen/src/test/java/io/harness/cdng/helm/HelmRollbackStepTest.java (hunk 7 above)
Comment (GoelPratyush99) [reply 125-cd-nextgen/src/test/java/io/harness/cdng/helm/HelmRollbackStepTest.java]: done

Comment (GoelPratyush99) [reply 125-cd-nextgen/src/test/java/io/harness/cdng/helm/HelmRollbackStepTest.java]: done

**Comment by GoelPratyush99** [comment]
trigger uts

**Review by TarunUba1 — APPROVED** [review]

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java [hunk 9]
```diff
@@ -539,8 +539,10 @@ private List<ContainerInfo> getKubectlContainerInfos(HelmCommandRequestNG comman
         if (!commandRequest.isSkipSteadyStateCheck()) {
           Optional<String> ocPath = setupPathOfOcBinaries(entry.getValue());
           ocPath.ifPresent(commandRequest::setOcPath);
+          List<KubernetesResourceId> kubernetesResourceIds =
+              filterKubernetesJobResource(entry.getValue(), commandRequest.isUseSteadyStateCheckForJobs());
```
[the remark this answers is not in the document]
Comment (GoelPratyush99) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/helm/HelmDeployServiceImplNG.java]: done

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger uts

**Review by TarunUba1 — APPROVED** [review]

**Review by markoharness — APPROVED** [review]

**Review by TarunUba1 — APPROVED** [review]

**Review by nishantsaini20 — APPROVED** [review]
Approved for ng-settings

**Comment by GoelPratyush99** [comment]
trigger review

**Pull request merged**